src/parser/de.rs

```diff
@@ -0,0 +1,303 @@
+                }
+            }
+        }
+        Ok(s.clone())
+    }
+}
+
+fn de_kebab(qry: &str) -> String {
+    let mut s = String::new();
+    let mut b = false;
+    for c in qry.chars() {
+        match c {
+            '-' => {
+                b = true;
+            }
+            _ => {
+                if b {
+                    s.push(c.to_ascii_uppercase());
+                }else {
+                    s.push(c);
+                }
+                b = false;
+            }
+        }
+    }
+    s
+}
+
+fn to_json(
+    value: &Value,
+    level: usize,
+    parent_key: Option<String>,
+    result: &mut HashMap<usize, HashMap<String, Vec<(String, String)>>>,
+    root_node: &String,
+) {
+    match value {
+        Value::Null | Value::Bool(_) | Value::Number(_) => (),
+        Value::String(s) => {
+            let y = parent_key.unwrap_or_default();
+            result
+                .entry(level)
+                .or_insert_with(HashMap::new)
+                .entry(root_node.clone())
+                .or_insert_with(Vec::new)
+                .push((y, s.clone()));
+        }
+        Value::Array(arr) => {
+            for v in arr.iter() {
+                to_json(v, level + 1, parent_key.clone(), result, root_node);
+            }
+        }
+        Value::Object(obj) => {
+            for (k, v) in obj.iter() {
+                to_json(v, level + 1, Some(k.to_string()), result,if v.is_object() {
+                    k
+                }else {
+                    root_node
+                });
+            }
+        }
+    }
+}
+
+fn to_json_str(value: &Value) -> String {
+    match value {
+        Value::Null => "".to_string(), // Return empty string for null values
+        Value::Bool(b) => b.to_string(),
+        Value::Number(num) => num.to_string(),
+        Value::String(s) => format!("{}", s),
+        Value::Array(arr) => {
+            let elements: Vec<String> = arr.iter().map(|v| to_json_str(v)).collect();
+            format!("[{}]", elements.join(" "))
+        }
+        Value::Object(obj) => {
+            let pairs: Vec<String> = obj.iter().map(|(k, v)| get_cur(k, v)).collect();
+            format!("{{{}}}", pairs.join(" "))
+        }
+    }
+}
+
+fn get_cur(k: &String, v: &Value) -> String {
+    let s = to_json_str(v);
+    if s.is_empty() {
+        k.clone()
+    } else {
+        format!("{} {}", k, s)
+    }
+}
+
+fn next_token(input: &str, position: &mut usize) -> Option<char> {
+    if let Some(ch) = input.chars().nth(*position) {
+        *position += 1;
+        Some(ch)
+    } else {
+        None
+    }
+}
+
+#[cfg(test)]
+mod de_tests {
+    use crate::async_graphql_hyper::GraphQLRequest;
+    use crate::parser::de::Parser;
+
+    #[test]
+    fn parse_t() {
+        let mut parser = Parser::from_qry("user&$m=id=123,address.city=foo&$s=name,age,address.city,address.state");
+        let x = parser.parse::<GraphQLRequest>();
+        println!("{:?}", x);
+    }
```

Comment:
Let's use a built in query parser - should be available in hyper or reqwest I think.